pageserver/src/walredo.rs

```diff
@@ -332,9 +334,17 @@ impl PostgresRedoManager {
             anyhow::bail!("invalid neon WAL redo request with no base image");
         }
 
-        // Apply all the WAL records in the batch
-        for (record_lsn, record) in records.iter() {
-            self.apply_record_neon(key, &mut page, *record_lsn, record)?;
+        // process the records in batches and yield; this should guard against pathological
+        // situations where we accidentially have a huge number of in-neon applied records.
+        let yield_every = 200;
+
+        for records in records.chunks(yield_every) {
```

Comment:
It would be nice to have some sort of generic yielding iteration construct that yields if it has hogged the thread for a certain amount of time. This works just as well though.

Reply:
I view it as rust code is fast, we are basically doing in-place mutation of an already allocated page. I would expect one iteration to be few nanos or micros, with this kind of limit we are quite close to millisecond worst case.

If we'd do timing at this level it might be too costly. Best would be https://docs.rs/tokio/latest/tokio/task/fn.consume_budget.html which would just decrement a task-local, but that is still unstable, sadly.

Reply:
Not sure if it's applicable here, but there is a `yielding_loop` helper that wraps the "yield every n iterations" case.